components/core/src/clp/Grep.cpp

```diff
@@ -1063,4 +1011,437 @@ size_t Grep::search(Query const& query, size_t limit, Archive& archive, File& co
 
     return num_matches;
 }
+
+void Grep::generate_query_substring_logtypes(
+        string& processed_search_string,
+        ByteLexer& lexer,
+        vector<std::set<QueryLogtype>>& query_substr_logtypes
+) {
+    // We need to differentiate between literal '*'/'?' and wildcards
+    std::vector<bool> is_greedy_wildcard;
+    std::vector<bool> is_non_greedy_wildcard;
+    std::vector<bool> is_cancel;
+    get_wildcard_and_cancel_locations(
+            processed_search_string,
+            is_greedy_wildcard,
+            is_non_greedy_wildcard,
+            is_cancel
+    );
+
+    // Consider each substr(j,i) of the processed_search_string and determine if it could have been
+    // compressed as static-text, a variable, or some combination of variables/static-text
+    // Then we populate each entry in query_substr_logtypes which corresponds to the logtype for
+    // substr(0,n). To do this, for each combination of substr(j,i) that reconstructs substr(0,n)
+    // (e.g., substring "*1 34", can be reconstructed from substrings "*1", " ", "34"), store all
+    // possible logtypes (e.g. "*<int> <int>, "*<has#> <int>, etc.) that are unique from any
+    // previously checked combination. Each entry in query_substr_logtypes is used to build the
+    // following entry, with the last entry having all possible logtypes for the full query itself.
+    bool i_is_cancelled = false;
+    for (uint32_t i = 0; i < processed_search_string.size(); i++) {
```

Comment:
* Let's rename `i` to `end_idx` for clarity.
* Let's make `i`/`end_idx` exclusive to match the behaviour of existing code.